pkg/syncer/storage/storage_pv_controller.go

```diff
@@ -0,0 +1,210 @@
+			key:       syncTargetKey,
+		},
+		updateDownstreamPersistentVolumeClaim: func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+			persistentVolumeClaim, err := downstreamKubeClient.CoreV1().PersistentVolumeClaims(persistentVolumeClaim.Namespace).Update(ctx, persistentVolumeClaim, metav1.UpdateOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			return persistentVolumeClaim, nil
+		},
+		getDownstreamPersistentVolumeClaim: func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error) {
+			return downstreamInformers.ForResource(persistentVolumeClaimSchemeGroupVersion).Lister().ByNamespace(persistentVolumeClaimNamespace).Get(persistentVolumeClaimName)
+		},
+		getUpstreamPersistentVolume: func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error) {
+			// We don't need the namespace since the PV is cluster scoped.
+			return upstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Lister().ByCluster(clusterName).Get(persistentVolumeName)
+		},
+		getDownstreamPersistentVolumeFromNamespaceLocator: func(namespaceLocator shared.NamespaceLocator) (runtime.Object, error) {
+			namespaceLocatorJSONBytes, err := json.Marshal(namespaceLocator)
+			if err != nil {
+				return nil, err
+			}
+			persistentVolumes, err := downstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Informer().GetIndexer().ByIndex(byNamespaceLocatorIndexName, string(namespaceLocatorJSONBytes))
+			if err != nil {
+				return nil, err
+			}
+			if len(persistentVolumes) == 0 {
+				return nil, nil
+			}
+			// There should be only one persistent volume with the same namespace locator, return it.
+			return persistentVolumes[0].(*unstructured.Unstructured), nil
+		},
+	}
+
+	logger := logging.WithReconciler(syncerLogger, persistentVolumeControllerName)
+
+	// Add the PV informers to the controller to react to upstream PV events.
+	logger.V(2).Info("Setting upstream up informer", persistentVolumeSchemeGroupVersion.String())
+	upstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.AddToQueue(obj, logger)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.AddToQueue(newObj, logger)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.AddToQueue(obj, logger)
+		},
+	})
+
+	// Add the downstream PV informer to the index
+	err := downstreamInformers.ForResource(persistentVolumeSchemeGroupVersion).Informer().AddIndexers(cache.Indexers{byNamespaceLocatorIndexName: shared.IndexByNamespaceLocator})
+	if err != nil {
+		return nil, err
+	}
+	c.downstreamInformer = downstreamInformers.ForResource(persistentVolumeSchemeGroupVersion)
+
+	return c, nil
+}
+
+func (c *PersistentVolumeController) AddToQueue(obj interface{}, logger logr.Logger) {
+	key, err := keyfunctions.DeletionHandlingMetaNamespaceKeyFunc(obj)
```

Comment:
If you are queueing objects from the upstream (kcp), you need the cluster-aware keyfunctions & splitters

Reply:
I assume you meant `kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc()` so I went ahead and used it. Let me know if I misunderstood. Thanks!